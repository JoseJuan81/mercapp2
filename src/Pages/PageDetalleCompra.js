import React from 'react'

export const PageDetalleCompra = () => {
    return (
        <div>
            Detalle de compra
        </div>
    )
}
