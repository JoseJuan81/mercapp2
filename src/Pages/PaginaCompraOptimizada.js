import { isEmpty } from 'functionallibrary';
import React, { useEffect, useState } from 'react';
import { useSelector } from 'react-redux';

import { calculateTotal } from '../helper/calculateTotal';
import { capitalizeText } from '../helper/capitalize';
import { clasifyingInsumosByEstablishment } from '../helper/clasifyingInsumosByEstablishment';

export const PaginaCompraOptimizada = () => {

    // ===== STORE =====
    const { insumos } = useSelector( store => store.newPurchase );

    // ===== STATE =====
    const [insumosByEstablishment, setInsumosByEstablishment] = useState([]);

    useEffect(() => {
        
        const insumosClasified = clasifyingInsumosByEstablishment( insumos );
        setInsumosByEstablishment( insumosClasified )

    }, [insumos]);

    return (
        <>
            {isEmpty( insumos )
                ? <NoInsumosSelected />
                : <InsumosByEstablishment insumos={ insumosByEstablishment } />
            }
        </>
    )
}

const InsumosByEstablishment = ({ insumos }) => {

    return (
        <div
            className="
                px-2
            "
        >
            {
                insumos.map(( { name, insumos }, index ) => (
                    <EstablishmentContainer
                        key={ `${ name } - ${ index }` }
                        name={ name }
                        insumos={ insumos }
                    />
                ))
            }
        </div>
    )
}

const EstablishmentContainer = ({ name, insumos }) => {

    // ===== STATE =====
    const [open, setOpen] = useState( false );

    // ===== VARIABLES LOCALES =====
    const qty = insumos.length;
    const total = calculateTotal( insumos, name );

    const establishmentName = capitalizeText( name );

    return (
        <div
            className="
                px-2 mb-2 pt-2
                border border-solid border-warmGray-300
                rounded
            "
            onClick={ () => setOpen( o => !o ) }
        >
            
            <InsumoContainerHeader
                quantity={ qty }
                total={ total }
                establishmentName={ establishmentName }
            />

            <InsumosList
                name={ name }
                insumos={ insumos }
                open={ open }
            />

        </div>
    )
}

const InsumoContainerHeader = ({ quantity, total, establishmentName}) => {
    return (
        <>
            <div
                className="
                    flex items-center justify-between
                    text-xl font-bold
                "
            >
                <h1
                    className="
                        text-warmGray-600 
                    "
                >{ establishmentName }</h1>
                <h3
                    className="
                        text-warmGray-800
                    "
                >{ total }</h3>
            </div>
            <span
                className="
                    text-warmGray-500
                "
            >Cantidad de insumos: { quantity }
            </span>
        </>
    )
}

const InsumosList = ({ insumos, open, name }) => {
    return (
        <ul
            className={`
                ${ open ? 'max-h-96' : 'max-h-0' }
                duration-200
                overflow-auto
                mt-2
            `}
        >
            {insumos.map(( insumo, i ) => (
                <li
                    key={`${insumos.name}-${i}`}
                    className={`
                        ${ i === 0 ? 'border-none' : 'border-t' }
                        border-solid border-warmGray-300
                        flex items-center justify-between
                        ${ i % 2 === 0 ? 'bg-white' : 'bg-warmGray-50' }
                    `}
                >
                    <h4
                        key={ insumo + i }
                    >{ insumo.name }
                    </h4>
                    <h5>
                        { insumo.price[name] }
                    </h5>
                </li>
            ))}
        </ul>
    )
}

const NoInsumosSelected = () => {
    return (
        <>
            <h1>No ha seleccionado insumos</h1>
            <p>Seleccione varios insumos y aqui vera donde tienen el mejor precio de compra</p>
        </>
    )
}
