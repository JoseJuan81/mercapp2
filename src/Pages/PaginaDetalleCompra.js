import { round } from 'functionallibrary';
import React from 'react';

import { getFormatDate } from '../helper/dates';
import { paretoOrdering } from '../helper/paretoOrdering';

const TWODECIMALS = round( 2 );

export const PaginaDetalleCompra = ({ details }) => {

    // ===== VARIABLES LOCALES =====
    const { closed, createdAt, establishmentName, insumos, name, total } = details;
    const insumosOrdered = paretoOrdering( insumos, { prop: 'total', total } );

    return (
        <div>
            {name && <h1>Nombre: { name }</h1>}
            {createdAt && <h1>fecha creacion: { getFormatDate( createdAt ) }</h1>}
            {establishmentName && <h1>comprado en: { establishmentName }</h1>}
            <h1>Insumos comprados: { insumos.length }</h1>
            <h1>gastado: S/. { total }</h1>
            <h1>estado de la compra: { closed ? 'Cerrada' : 'Abierta' }</h1>
            <InsumosList insumos={ insumosOrdered } />
        </div>
    )
}

const InsumosList = ({ insumos }) => {

    const tableClassRow = `
        grid grid-cols-5
        items-center justify-center
        border-b border-solid border-warmGray-200
    `;
    const classCell = `
        h-full
        flex items-center
        px-1 py-2
    `;

    return (
        <ul
            className="px-2"
        >
                <li
                    className={ tableClassRow }
                >
                    <span
                        className="
                            col-span-2
                            border-l border-solid border-warmGray-200
                        "
                    >Nombre</span>
                    <span
                        className="text-center"
                    >Precio</span>
                    <span
                        className="text-center"
                    >Cant.</span>
                    <span
                        className="text-center"
                    >Total</span>
                </li>
            {insumos.map((insumo, index) => (
                <li
                    key={ insumo.id + index }
                    className={ tableClassRow }
                >
                    <div
                        className={`
                            ${ classCell }
                            col-span-2
                            justify-start
                        `}
                    >{ insumo.name }</div>
                    <div
                        className={`
                            ${ classCell }
                            justify-center
                        `}
                    >{ TWODECIMALS( insumo.total / insumo.quantity ) }</div>
                    <div
                        className={`
                            ${ classCell }
                            justify-center
                        `}
                    >{ insumo.quantity }</div>
                    <div
                        className={`
                            ${ classCell }
                            justify-center
                        `}
                    >{ insumo.total }</div>
                    <div
                        className={`
                            ${ classCell }
                            justify-center
                            col-span-2
                        `}
                    >
                        <span>peso: </span>
                        <span>{ insumo.weight }%</span>
                    </div>
                    <div
                        className={`
                            ${ classCell }
                            justify-center
                            col-span-3
                        `}
                    >
                        <span>acumulado: </span>
                        <span>{ insumo.weightAcc }%</span>
                    </div>
                </li>
            ))}
        </ul>
    )
}
