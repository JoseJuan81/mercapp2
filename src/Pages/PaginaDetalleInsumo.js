import { isEmpty } from 'functionallibrary';
import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { startLoadingInsumoData } from '../actions/insumoDetailsAction';
import { InsumoDataInfo } from '../components/Insumos/InsumoDataInfo';
import { InsumoPriceInfo } from '../components/Insumos/InsumoPriceInfo';
import { InsumoStatisticsInfo } from '../components/Insumos/InsumoStatisticsInfo';

export const PaginaDetalleInsumo = () => {
    
    // ===== NAVIGATION =====
    const url = new URL( window.location );
    const insumoId = url.searchParams.get( 'insumoId' );


    // ===== STORE =====
    const dispatch = useDispatch()
    const { data, prices } = useSelector( store => store.insumoDetails );
    const { loading } = useSelector( store => store.loading );

    // cargar datos del insumo
    useEffect(() => {

        if ( insumoId !== data.id ) {

            dispatch( startLoadingInsumoData( insumoId ) );
        }

    }, [insumoId, data.id, dispatch]);

    return (
        <div
            className="
                pt-3
                flex items-start justify-center
            "
        >

            {!isEmpty( url.searchParams.get('data') ) && (
                <InsumoDataInfo data={ data } prices={ prices } />
            )}

            {!isEmpty( url.searchParams.get('statistics') ) && !loading && (
                <InsumoStatisticsInfo historyPrices={ data.historyPrice } name={ data.name } />
            )}

        </div>
    )
}
