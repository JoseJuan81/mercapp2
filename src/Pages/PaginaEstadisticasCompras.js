import { isEmpty, round } from 'functionallibrary';
import React, { useEffect, useState } from 'react';

import ParetoBarChart from '../helper/paretoInsumosChart';

import { extractAndOrderInsumos } from '../helper/extractAndOrderInsumos';
import { getTotalAmountAndTotalInsumos } from '../helper/getTotalAmountAndTotalInsumos';

const TWODECIMALS = round( 2 );

export const PaginaEstadisticasCompras = ({ purchases, currency = 'S/.' }) => {

    // ===== STATE =====
    const [width, setWidth] = useState( window.innerWidth );

    // ===== VARIABLES LOCALES =====
    const { totalInsumos, totalAmount } = getTotalAmountAndTotalInsumos( purchases );
    const {
        insumos:allInsumosOrdered,
        expensive,
        expense
    } = extractAndOrderInsumos( purchases );


    // ===== FUNCIONES =====
    const getScreenWidth = () => {
        
        setWidth( window.innerWidth );
    }

    // ESCUCHAR CAMBIOS DE DIMENSION EN PANTALLA
    useEffect(() => {
        window.addEventListener( 'resize', getScreenWidth )
        return () => {
            window.removeEventListener( 'resize', getScreenWidth )
        }
    }, [width])

    // CREAR GRAFICO EN FUNCION DEL ANCHO DE LA PANTALLA
    useEffect(() => {

        if ( !isEmpty( allInsumosOrdered ) ) {

            const container = document.querySelector('#d3-pareto__insumos');

            ParetoBarChart.init( container, {
                data: allInsumosOrdered,
                prop: 'total',
                margin: { top: 20, right: 15, bottom: 10, left: 25 },
                width
            });
        }

    }, [width, allInsumosOrdered])

    return (
        <div>

            <div
                className="
                    my-4 px-2 mx-auto
                    grid grid-cols-2 gap-3
                    text-center
                    max-w-md
                "
            >
                <CardDescriptor
                    label="Compras"
                    data={ purchases.length }
                />
                <CardDescriptor
                    userClass="animate__delay-1s"
                    label="Insumos" data={ totalInsumos }
                />
                <CardDescriptor
                    userClass="animate__delay-2s col-span-2"
                    label={ `Total ${currency}` }
                    data={ totalAmount }
                />
                <CardExpensive
                    userClass="animate__delay-3s col-span-2"
                    label="Mas costoso"
                    name={ expensive.name }
                    amount={ TWODECIMALS(expensive.total / expensive.quantity ) }
                />
                <CardExpensive
                    userClass="animate__delay-4s col-span-2"
                    label="Gaste mas"
                    name={ expense.name }
                    amount={ TWODECIMALS( expense.total ) }
                />
                
            </div>

            <div id="d3-pareto__insumos" className="pb-5"></div>
        </div>
    )
}

const CardDescriptor = ({ label, data, userClass = "" }) => {

    return (
        <div
            className={`
                ${ userClass }
                animate__animated animate__bounceInLeft
                border border-solid border-warmGray-200 rounded-lg
                pl-2 pr-4 py-1
                flex items-end justify-between
                h-12
            `}
        >
            <label
                className="
                    text-warmGray-400 font-bold text-sm text-left
                "
            >{ label }</label>
            <span
                className="
                    text-lime-500 font-bold text-4xl text-right
                    transform translate-y-1
                "
            >{ data }</span>
        </div>
    )
}

const CardExpensive = ({ label, name, amount, userClass }) => {

    return (
        <div
            className={`
                ${ userClass }
                animate__animated animate__bounceInLeft
                border border-solid border-warmGray-200 rounded-lg
                pl-2 pr-4 py-2
                flex flex-col
                relative
            `}
        >
            <span
                className="
                text-warmGray-500 font-bold text-xl text-left
                "
            >{ name }</span>
            <div className="flex items-end justify-between">
                <label
                    className="
                        text-rose-500 font-bold text-lg text-left
                    "
                >{ label }</label>
                <span
                    className="
                        text-rose-500 font-bold text-4xl text-right
                        bg-rose-50
                        rounded
                        px-3 py-1
                    "
                >
                    { amount }
                </span>
            </div>

            <i
                class="
                    fas fa-exclamation-circle
                    text-rose-700
                    absolute -top-2 -left-1
                "
            ></i>

        </div>
    )
}
