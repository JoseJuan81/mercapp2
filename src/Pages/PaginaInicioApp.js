import { map, upperFirst } from 'lodash';
import React, { useState } from 'react';
import { useSelector } from 'react-redux';

import { BackButton } from '../components/Buttons/AppButtons';
import { AreaChart } from '../components/d3/AreaChart';
import { ProgressBar } from '../components/d3/ProgressBar';

import { getMonthInWord } from '../helper/dates';
import { getLastMonths } from '../helper/getLastMonths';
import { getTotalByMonth } from '../helper/totalByMonth';

export const PaginaInicioApp = () => {

	// === STORE ===
	const { expenses, currencies } = useSelector( store => store.user );

	// === STATE ===
	const [currentMonth, setCurrentMonth] = useState( new Date().getMonth() );

	// === VARIABLES LOCALES ===
	const monthInWords = getMonthInWord( new Date().setMonth( currentMonth ) );
	const totalByMonth = getTotalByMonth( expenses );
	const last3ExpenseMonth = getLastMonths({ period: 3, data: totalByMonth, currentMonth });

	const [currency] = currencies;

	return (
		<div
			className=""
		>
			<div
				className="
					flex justify-center
				"
			>
				<BackButton
					isButton
					className="w-10"
					onClick={ () => setCurrentMonth( m => m - 1 ) }
				/>

				<MonthlyExpensesCard
					monthDate={ monthInWords }
					amount={ totalByMonth[currentMonth]?.total || 0 }
					currency={ currency?.symbol }
					diferencePercentageToPrevius={ 9.89 }
					last3ExpenseMonth={ last3ExpenseMonth }
				/>

				<BackButton
					isButton
					className="w-10 transform rotate-180"
					onClick={ () => setCurrentMonth( m => m + 1 ) }
				/>
			</div>
			
			<CategorySummaryCard
				currentMonth={ totalByMonth[currentMonth] }
				currency={ currency?.symbol }
			/>

			<EstablishmentSummaryCard
				currentMonth={ totalByMonth[currentMonth] }
				currency={ currency?.symbol }
			/>
			
		</div>
	)
}

export const MonthlyExpensesCard = ({
	monthDate,
	amount,
	currency,
	diferencePercentageToPrevius,
	last3ExpenseMonth,
}) => {
	return (
		<div
			className="
				rounded-lg shadow-md
				mt-4 mb-8 px-4 pt-4 pb-6	
				max-w-60
			"
			>
			<div
				className="
					flex flex-col items-center
				"
			>
				<MonthDate monthDate={ monthDate } />

				<MonthAmount
					currency={ currency }
					amount={ amount }
					diferencePercentageToPrevius={ diferencePercentageToPrevius }
				/>

				<AreaChart
					lastAmountsByMonth={ last3ExpenseMonth }
				/>

			</div>
		</div>
	)
}

export const MonthDate = ({ monthDate }) => {
	return (
		<span className="text-lime-500 font-bold">{ monthDate }</span>
	)
}

export const MonthAmount = ({ currency, amount, diferencePercentageToPrevius }) => {
	return (
		<div
			className="
				flex items-center justify-center
				my-4
				w-full
				relative
			"
		>
			<div
				className="
						text-5xl font-bold text-warmGray-700
						flex items-end justify-center
						col-span-3
				"
			>
				<span
					className="
						text-sm text-warmGray-500
						mr-2
					"
				>{ currency }</span>
				<span>{ amount }</span>
			</div>

			{/* <CompareToPrevius diferencePercentageToPrevius={ diferencePercentageToPrevius } /> */}
			
		</div>
	)
}

export const CompareToPrevius = ({ diferencePercentageToPrevius }) => {
	return (
		<div
			className="
				absolute top-0 left-0
				w-full h-full
				flex justify-end items-center
			"
		>
			<div className="w-auto h-10 flex items-center">
				<i className="fas fa-arrow-circle-up"></i>
			</div>
			<div>{ diferencePercentageToPrevius }</div>
			<span>%</span>
		</div>
	)
}

export const CategorySummaryCard = ({ currentMonth, currency }) => {
	return (
		<div
			className="
				mx-6 px-4 py-4 mb-8
				flex flex-col items-center
				rounded-lg
				shadow-md
			"
		>
			<h2
				className="
					text-lg font-bold text-sky-500
				"
			>Acumulado por categor√≠a</h2>
			<ul
				className="
					divide-y-2 divide-warmGray-100
					mt-2
				"
			>
				{ map( currentMonth?.totalByCategory, (val, key) => (
					<li
						key={ key }
						className="
							grid grid-cols-6
							items-center
							h-10
							font-semibold text-sm
							pt-3 mb-3
						"
					>
						<span 
							className="
								col-span-2
								font-semibold text-sm text-warmGray-600
							"
						>{ upperFirst( key ) }</span>

						<ProgressBar
							className="
								col-span-3
								pl-2 pr-3 -ml-2
							"
							percentage={ val.percentage }
						/>
						
						<span
							className="
								justify-self-end
								text-base text-warmGray-600
							"
						>
							<span className="text-xs mr-1">{ currency }</span>
							{ val.total }
						</span>
					</li>
				)) }

			</ul>
		</div>
	)
}

export const EstablishmentSummaryCard = ({ currentMonth, currency }) => {
	return (
		<div
			className="
				mx-6 px-4 py-4 mb-8
				flex flex-col items-center
				rounded-lg
				shadow-md
			"
		>
			<h2
				className="
					text-lg font-bold text-lime-500
				"
			>Acumulado por establecimiento</h2>
			<ul
				className="
					divide-y-2 divide-warmGray-100
					mt-2
				"
			>
				{ map( currentMonth?.totalByEstablishment, (val, key) => (
					<li
						key={ key }
						className="
							grid grid-cols-6
							items-center
							h-10
							font-semibold text-sm
							pt-3 mb-3
						"
					>
						<span 
							className="
								col-span-2
								font-semibold text-sm text-warmGray-600
							"
						>{ upperFirst( key ) }</span>

						<ProgressBar
							className="col-span-3 px-2"
							percentage={ val.percentage }
						/>
						
						<span
							className="
								justify-self-end
								text-base text-warmGray-600
							"
						>
							<span className="text-xs mr-1">{ currency }</span>
							{ val.total }
						</span>
					</li>
				)) }

			</ul>
		</div>
	)
}
