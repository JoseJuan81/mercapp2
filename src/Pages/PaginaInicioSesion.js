import React, { useContext } from 'react';
import { Link, useHistory } from 'react-router-dom';

import { misInsumosPath, registroUsuarioPath } from '../constant/routes';
import { user } from '../constant/user';
import { UserContext } from '../context/User/UserContext';
import { useForm } from '../hooks/useForm';

export const PaginaInicioSesion = () => {

    const { dispatch } = useContext( UserContext );

    const history = useHistory();

    const { formState, handleInputChange, invalidForm } = useForm({
        email: '',
        password: ''
    }, ['email', 'password']);

    const handleLogIn = (ev) => {
        ev.preventDefault();
        
        dispatch({ type: user.login, payload: formState });
        history.replace( misInsumosPath );
    }

    return (
        <div
            className="
                flex items-center justify-center
                h-screen
            "
        >
            <form
                className="
                    flex flex-col
                    py-6 px-4 mx-4
                    w-full
                    border border-solid border-warmGray-200
                    rounded-xl shadow-lg
                "
                onSubmit={ handleLogIn }
            >
                <h1
                    className="
                        text-4xl font-semibold text-lime-500 text-center
                        pb-4 mb-2
                    "
                >
                    MercApp 2
                </h1>

                <input
                    required
                    className="
                        w-full h-16
                        border border-solid border-warmGray-200
                        rounded
                        pl-4 pr-2 mb-4
                    "
                    placeholder="Introduzca su correo"
                    type="email"
                    name="email"
                    value={ formState.email }
                    onChange={ handleInputChange }
                />
                <input
                    required
                    className="
                        w-full h-16
                        border border-solid border-warmGray-200
                        rounded
                        pl-4 pr-2 mb-4
                    "
                    placeholder="Introduzca su contrasena"
                    type="password"
                    name="password"
                    maxLength={ 16 }
                    minLength={ 6 }
                    value={ formState.password }
                    onChange={ handleInputChange }
                />
                <button
                    disabled={ invalidForm }
                    className={`
                        w-full h-16
                        bg-lime-200
                        text-lime-700 text-xl
                        rounded
                        ${invalidForm && 'opacity-30'}
                    `}
                >
                    Inicia sesion
                </button>

                <Link
                    className="
                        bg-warmGray-100
                        text-center text-warmGray-700
                        py-2 mt-4
                    "
                    to={ registroUsuarioPath }
                >Registrarme</Link>
            </form>

        </div>
    )
}
