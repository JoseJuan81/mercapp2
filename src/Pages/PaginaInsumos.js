import React from 'react';

import { PageTitle } from '../components/Genericos/PageTitle';
import { InsumosContainer } from '../components/container/InsumosContainer';


export const PaginaInsumos = () => {

    return (
        <div className="px-2 pb-20 mt-4 relative">
            
            <PageTitle title="Insumos" />

            <InsumosContainer />
            
        </div>
    )
}
