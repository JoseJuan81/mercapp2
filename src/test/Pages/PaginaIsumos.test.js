import { shallow } from 'enzyme';

import { PaginaInsumos } from '../../Pages/PaginaInsumos';

describe('Prueba sobre archivo PaginaInsumo', () => {

    test('Se debe renderizar correctamente', () => {

        const wrapper = shallow( <PaginaInsumos />);

        expect( wrapper ).toMatchSnapshot();
    })
})