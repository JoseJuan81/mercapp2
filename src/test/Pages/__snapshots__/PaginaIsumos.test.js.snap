// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba sobre archivo PaginaInsumo Se debe renderizar correctamente 1`] = `
<div
  className="px-2 pb-20 mt-4 relative"
>
  <Memo()
    title="Insumos"
  />
  <InsumosContainer />
</div>
`;
